### ISSUE TITLE
Provider crashes when applying an ICSP configuration without an ICSP endpoint defined at the provider level

### ISSUE THREAD

**Issue by przemolb** [HewlettPackard/terraform-provider-oneview#40]
Hello,

I am trying to use terraform v0.9.6 to provision some test servers with Oneview. Terraform is crashing  when it comes to ICSP definition. Below is all the config files, debug and output:

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.

provider "oneview" {
  ov_username   = "user"
  ov_password   = "pass"
  ov_endpoint   = "https://oneview-test.some.domain/"
  ov_sslverify  = false
  ov_apiversion = 200
}

resource "oneview_server_profile_template" "default" {
  name = "terraform-test-server-profile-template"
  enclosure_group = "EG-LAB"
  server_hardware_type = "BL460c Gen8 2"
}

resource "oneview_server_profile" "default" {
  name = "terraform-test-server-profile"
  hardware_name = "PENUK2HLAB001, bay 9"
  hw_filter = [ "position = 9" ]
  template = "${oneview_server_profile_template.default.name}"
  type = "ServerProfileTemplateV5"
  power_state = "on"
}

resource "oneview_icsp_server" "default" {
  ilo_ip = "x.x.x.x"
  user_name = "some_user"
  password = "some_pass"
  serial_number = "${oneview_server_profile.default.serial_number}"
  build_plans = ["terraform-rhel-7.2-x64-scripted-install"]
}
```
There is also debug output:
https://gist.github.com/przemolb/f8510e7590ac1b91ca6f8b124844f942

and panic output
https://gist.github.com/przemolb/7debb7b85e2bfd6405c3b1b1e944d0b3

**Comment by patrickdappollonio** [comment]
Thank you so much for the so-detailed bug! It seems that there's a check missing, but it may also have some relationship with [oneview-golang](https://github.com/HewlettPackard/oneview-golang). I'll check it out and provide a fix asap!

**Comment by patrickdappollonio** [comment]
TL;DR: You're missing some configs in your provider config, specifically: `icsp_endpoint`, `icsp_domain`, `icsp_username`, `icsp_password`, `icsp_endpoint`, `icsp_sslverify`, `icsp_apiversion`.

Long story: This is a shared responsibility between [`oneview-golang`](https://github.com/HewlettPackard/oneview-golang) (pinging @wenlock for this) and us. Due to a base decision on how the SDK for Go will work, it allows shadowing since, to create an object, you need a previous, `nil` (empty) one to make it work. This gives these kind of nasty bugs that are difficult to troubleshoot.

Terraform provider OneView has a configuration rule that checks if `icsp_endpoint` is set, if so, [it blindly passes all ICSP parameters back to the original `api` structure](https://github.com/HewlettPackard/terraform-provider-oneview/blob/6e53077ba92a1f3e4ec39ead3db1dcad88111792/oneview/provider.go#L124-L135) which contains inside either a pointer to the ICSP client, or a pointer to nowhere. In this case, since there was no `icsp_endpoint` in your provider configuration, then that statement got bypassed, creating no ICSP client. Later on, the code tries to call ICSP actions against a nil ICSP, which results in the crash you see. 

Unfortunately, there's no "correct" easy fix -- although there's an easy fix I'll be providing in a minute -- since we will have to modify `oneview-golang` so the actions have a pre-sanity check to check for initialization. If it never happened, then it should never continue. It's also our provider's fault though, since because [we're embedding 3 objects into the configuration](https://github.com/HewlettPackard/terraform-provider-oneview/blob/6e53077ba92a1f3e4ec39ead3db1dcad88111792/oneview/config.go#L39-L41), then there's no guarantees that when we call one of them it'll be set. The embedding isn't the best solution, but it works in the short-term.

I'll send a temporary workaround that'll report back if the ICSP settings were not set and someone was calling ICSP actions, and then I'll contact @wenlock and some others within our company to see what to do next.

Thanks for the submission!

**Comment by przemolb** [comment]
Hi @patrickdappollonio ,

thanks for firm and detailed description.
The configuration I have used is based on https://github.com/HewlettPackard/terraform-provider-oneview/blob/master/docs/index.html.markdown where all the icsp_* settings are optional.
Maybe the documentation should be changed as well to make people aware that it might happen.

**Comment by patrickdappollonio** [comment]
I agree, we should make it more clear that by using ICSP you should at least provide an endpoint. All the other parameters, like you said, are optional, so that continues to be true, since they depend on your special configuration. 

The docs in the homepage are also "problematic" since they have an ICSP example but no mention of the ICSP server endpoint. 

We will fix first the code issue [on PR #41](https://github.com/HewlettPackard/terraform-provider-oneview/pull/41) and then we will come back for the docs.

Thanks!

**Comment by przemolb** [comment]
I have another bug with similar configuration but the error message doesn't tell me what is wrong (at least I am not able to figure it out). Do you want me to put the details here or raise another issue ?

**Comment by patrickdappollonio** [comment]
Please create a new issue. If we consider it'll be the same one we can close it as duplicated, but it's better to have each bug with its own track. Thanks!
